src/taskgraph/config.py

```diff
@@ -8,107 +8,89 @@
 import sys
 from dataclasses import dataclass
 from pathlib import Path
-from typing import Dict
+from typing import Any, Dict, List, Literal, Optional, Union
 
-from voluptuous import ALLOW_EXTRA, All, Any, Extra, Length, Optional, Required
+import msgspec
 
-from .util.caches import CACHES
 from .util.python_path import find_object
-from .util.schema import Schema, optionally_keyed_by, validate_schema
+from .util.schema import validate_schema
 from .util.vcs import get_repository
 from .util.yaml import load_yaml
 
 logger = logging.getLogger(__name__)
 
 
-#: Schema for the graph config
-graph_config_schema = Schema(
-    {
-        # The trust-domain for this graph.
-        # (See https://firefox-source-docs.mozilla.org/taskcluster/taskcluster/taskgraph.html#taskgraph-trust-domain)  # noqa
-        Required("trust-domain"): str,
-        Optional(
-            "docker-image-kind",
-            description="Name of the docker image kind (default: docker-image)",
-        ): str,
-        Required("task-priority"): optionally_keyed_by(
```

Comment:
I think we'll need to preserve the `optionally_keyed_by` function.

I can help you with this later, but at a high level, I think we should be able to dynamically return the proper Python type annotation for it.

Reply:
To illustrate what `optionally_keyed_by` is doing, it means in this case the following inputs would all be valid:

```
task-priority:
    by-project:
        autoland: high
        default: low
```
```
task-priority:
    by-level:
        3: high
        1: low
```
```  
task-priority: high
```

So we need to come up with a standard Python type annotation that captures this.. It might be a bit ugly, but I think it should be do-able.